### PR TITLE
Project restructure and merge http-server into lobby

### PR DESCRIPTION
**Pull request by DanVanAtta**
## Overview
This is a forward looking update that creates a number of new sub-projects, not yet all implemented.
This gives us a placeholder for moving files around as part of future updates.

Important in this change the http-server capability is broken, and grafted onto the lobby project.
Next iterations will work to launch the http-server when the lobby is launched.

### New Sub-project summary:
- lobby-client-java: this is a java socket client, existing code would be refactored into this project, largely from Messengers
- lobby-client-http: this is the http-client project renamed and is largely unchanged
- java-sockets: forward looking sub-project that will contain common java-socket code
- game-headed-client-java: forward looking sub-project to have java client code specific for interacting with headless game server.
- lobby-db-code: forward looking sub-project that will contain the DB layer of the lobby.
- lobby-engine: this layer is meant to help us keep network transport code split from logic code. It is a bit forward looking and
    ideally large sections of 'lobby' would be merged into this sub-projectg, as well as any other lobby-specific code remaining in game-core.

## Functional Changes
- Breaks http server deployment, will follow-up in future updates.

## Manual Testing
- Launched a headed game client, a local lobby and local bot.

## Additional Review Notes

Perhaps the best summary is to look at the sub-project listing:
https://github.com/triplea-game/triplea/compare/master...DanVanAtta:lobby_work?expand=1#diff-88b7c47e47b8ee65263b6784b86fa0a7

Beyond that, as far as code, http-server was merged into lobby and http-client was renamed. As part of merging http-server some code landed in lobby-engine The rest of the updates are shell-projects for the "forward-looking" sub-projects

### CODE REVIEW

**Comment by DanVanAtta** [comment]
> I'm alright with these restructuring changes though want to make sure we don't go too far and end up with too many very small projects.

Good concern, I've been thinking about how far on the other spectrum is too much. On the one hand, I've thought it seems like a lot right now as to start we create these new code buckets and then slowly start to break things up. In the start you have a most of the new buckets and they are mostly empty; I'm hoping the rate of growth of new sub-projects will stabilize over time and increase much more slowly after these initial few break-ups.

I'm also leaning a bit more in the break things up into smaller pieces to start so we can get more code decoupling. Overly coupled code is a severe problem, going a bit too far into the decouple direction IMO is okay if that helps us get things cleaned up. It's quite easy to merge up sub-projects together.

Last thought I had on sub-projects, I think they should mostly map to the map code domains. I think there will be some more break-up of `game-core`, IMO I'd like that to be along game-rules, map parsing and perhaps save-game lines. In the end I'm hoping for sub-projects to help represent the major project domains.

**Comment by DanVanAtta** [comment]
> What is the difference between 'lobby-db' and 'lobby-db-code'?

I was initially wanting 'lobby-db' to be renamed to 'lobby-db-schema' or 'lobby-db-flyway', and then use 'lobby-db' for the java code. To avoid renaming packages, I opted to add one. I was also tempted to merge the two, but that got messy. We create a zip file of 'lobby-db' to be downloaded for flyway migrations, adding unnecessary library code to that is pretty nasty.


To answer the question more directly, lobby-db-code is in the diagram from this issue:
https://github.com/triplea-game/triplea/issues/4701

That diagram is only slightly out of date as the lobby and http-server are getting merged up. 'lobby-db-code' is a java library sub-project that is just the DB layer of the lobby. This is to help enforce a very hard API line and avoid any dependencies on from lobby to leak into the database layer. Of note, this was going to be more useful when http-server had a shared back-end, more important in that case. Regardless, it still seems very worthwhile. For example that module will have DB level integration testing and the others will not. It'll make it pretty clear that if you're doing DB testing in the 'lobby' project that something then has gone wrong. It'll also speed up test verifications when changing code in other sub-projects. Generally the isolation of code that should not change too much, tested in isolation, and is an API that we generally will want ot mock on its own, are all good things. For example we had examples where DAO concrete classes were being used instead of the interface, controllers that had no DAO. 



> What's the thinking behind having 'lobby-client-http' and 'lobby-client-java' separate?

They're very different dependencies/code-bases and should not have much in common. One is request/response objects and http endpoints, the other should be most of our 'MessengerChannel' code. These two packages should isolate as well any of the shared dependencies between game socket server and socket client. From a testing perspective these are also good system boundary locations. The rest of game-core when using lobby-client-java will no longer need an actual server, we can stub/mock the lobby-client-java API boundary. In turn, we'll have some live system testing to verify our clients work properly. For example, lobby-client-http is tested with wiremock to verify the client works correctly with an http server. In turn, http-server is launched in test and we use teh lobby-client-http to test that the endpoints work correctly. Thus, if anything correctly uses lobby-client-http, we know it should go on to work with the http-server. lobby-client-java is a step in this direction. Unfortunately that client API will be a bit difficult to pull out as it's mixed up with the API to talk to headless-bot socket and lobby server socket.



> Should project names have the technology/implementation in the name or just describe the purpose (avoid things like java/http/code)?


As long as we don't have a `triplea-common`, or a `java-util`, or a `java-common`, I'm happy with most names that we choose : )

In part the technology is important as it dictates the jar dependencies. If we wanted another client to the http-server, we would not necessarily want to include unused java-socket network utilities. Now that the http-server and lobby are being merged, the distinction is less, but the clients are tested very differently and have very different dependencies.

The packages with `java` in them are meant to indicate they are language level libraries. I'm really trying to help avoid the pitfall where we have a utility library that gets too large and then starts to accumulate business level logic accidently (for example, an analogy could be how Swing code code invaded all the rest of the TripleA logic code).

--------------

Long answers.. sorry for not cleanly editing them. Happy to discuss pros/cons further; but I would caution against analysis-paralysis at same time, i would like to integrate these changes and build on them further.

**Comment by ron-murhammer** [comment]
@DanVanAtta Ok. I think the separations make sense then. I'm still not in love with the names as when I see 'lobby-db' and 'lobby-db-code', it isn't very clear what the purpose of each of them is until you explained it. Maybe something more like 'lobby-db-schema' and 'lobby-db-persistence'. I think similarly when I see 'lobby-client-http' and 'lobby-client-java', I immediately think ok so the first must do some sort of HTTP type communication between client and lobby but the second I really have no idea as 'java' doesn't really mean anything IMO given that all the code base is written in java (if you actually had different language clients like java vs javascript vs C# then it might make more sense).

Mostly I'm just trying to see if there are potentially names that describe the function or purpose of the projects more clearly and avoid labeling it as 'code' or 'java' which I don't think means much in some of these cases. That being said, I don't plan to hold up this PR based on that just throwing out some ideas. I would like @RoiEXLab or @ssoloff to glance it over since we want to be in general agreement on these types of changes to avoid restructuring again and again.

**Comment by DanVanAtta** [comment]
Any thoughts to either of these:
(1) `lobby-db-schema` + `lobby-db-dao`
(2) `lobby-db-schema` + `lobby-db-code`

My default reaction is to leave existing names as-is when having trouble to decide. Hence 'lobby-db' remained.

'persistence' I do like on the one hand as it's a pretty well known concept from Spring and describe the code layer. On the other hand I do not like the concept of 'persistence' as being an object-store. For example, we can get the asnwer of  "is this person a moderator" from a DB query, we do not get it by loading a user object and checking the 'isAdmin()' method. Perhaps there is too much bias in my perspective, I strongly do not want to invite that kind of representation of the DB layer.

---------

For client libs, agree that 'java' is not that solid of a name in comparison to other lanugages.

What do you think of:
(1) lobby-socket-client + lobby-http-client

**Comment by ron-murhammer** [comment]
@DanVanAtta I like these ones:

**lobby-db-schema + lobby-db-dao** - makes it clear what each of these projects is for

**lobby-socket-client + lobby-http-client** or **lobby-client-socket + lobby-client-http** - The first reads more naturally though the second makes it so alphabetically those 2 would always appear together and both are clearly different lobby clients. Fine with either though.

File: lobby/README.md [hunk 1]
```diff
@@ -22,3 +22,54 @@ This is the well-known running TripleA lobby server.
...
+TODO: update below, from http-server component merge into lobby.
+
+## Overview
+
+This project is a server side component, a stand-alone http(s) server.
+
+### Functionality Provided / Strategic fit in the TripleA project
+
+We envision two kinds of client-server communication, over http or over websocket.
+Currently, lobby-server, uses Java RMI, which we would like to deprecate as it
```
Comment (RoiEXLab) [inline lobby/README.md]: This came up a lot of time.
Techincally we're not using RMI, but something very similar. It's probably good to explicitly state that here in the README

File: lobby/build.gradle [hunk 2]
```diff
@@ -11,12 +11,18 @@ mainClassName = 'org.triplea.lobby.server.LobbyRunner'
...
     implementation "org.mindrot:jbcrypt:$jbcryptVersion"
+    implementation 'com.google.code.gson:gson:2.8.5'
+    implementation 'com.sparkjava:spark-core:2.8.0'
 
     runtimeOnly "org.postgresql:postgresql:$postgresqlVersion"
+    runtimeOnly 'org.slf4j:slf4j-simple:1.7.21'
 
     testImplementation project(':test-common')
     testImplementation 'com.github.database-rider:rider-junit5:1.5.2'
+    testImplementation 'uk.co.datumedge:hamcrest-json:0.2'
```
Comment (RoiEXLab) [inline lobby/build.gradle]: Sorry if this seems obvious, but what exactly is this dependency for?

Comment (RoiEXLab) [reply lobby/build.gradle]: Ah, nevermind I got it.

**Comment by DanVanAtta** [comment]
@ron-murhammer I added a TODO comment to lobby-db/README.md that we would like to rename it. Changing the name of that project is not necessarily trivial as we do have other dependencies on that name. 

@RoiEXLab took a quick stab at some updates to the lobby/README.md. The document could use some help, for now all I can spare is the quick cleanup. Given there will be more updates coming, investing a lot of effort in good documentation that is non-redundant, general but still useful and not likely to go-out of date, IMO is not a terribly investment right now.

**Comment by RoiEXLab** [comment]
@DanVanAtta I'm fine with the documentation update. I agree that it is far from being perfect, but stating that it isn't exactly RMI is already enough for me.
Any chance you had a look at nested subprojects I mentioned?

**Comment by DanVanAtta** [comment]
I quickly get into analysis-paralysis when considering further sub-divisions. Will future http clients be more similar and should be grouped together, or will the lobby clients be more similar and grouped together?

That makes me think some organization is good, trying to get it perfect is overboard as it will change. Also to consider, any decision we make now is with less information than anything we can wait on, decisions made now are *worse* informed than decisions made later. 

I'm inclined for us to first focus on the bigger splits in the code. I suspect we do not have enough sub-projects of a same type that really need further sub-divisions. The original game-architecture, if you look to the oldest "core" of code you can see the building architecture was a 'server-client' model. That still will likely be the most natural split of code in `game-core`. I bring that up as the goal is to create fracture lines in the code so that inappropriate dependencies that reach across boundaries become disallowed. In the end is code that is more modular, where a given cluster of classes have a limited public APIs and any updates within that cluster are self-contained (and therefore more easily understand, updated and tested).

**Comment by DanVanAtta** [comment]
@RoiEXLab if you have some good ideas, it's certainly worth while to list/propose them for the sake of some consideration Perhaps later those ideas will become clearly the right fit.

**Pull request merged**